File: apps/website/internal/handlers/wiki.go
package handlers

import (
	"errors"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"github.com/tbd-milsim/reforger-backend/internal/middleware"
	"github.com/tbd-milsim/reforger-backend/internal/models"
)

// ListWiki returns the SOP nav list ordered for the left sidebar. The frontend
// groups by category.
//
// @route GET /api/v1/wiki
func (h *Handler) ListWiki(c *gin.Context) {
	var pages []models.WikiPage
	if err := h.db.Order("nav_order ASC").Order("title ASC").Find(&pages).Error; err != nil {
		logHandlerErr(c, "ListWiki", http.StatusInternalServerError, "could not list wiki")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not list wiki"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": pages})
}

// GetWikiPage returns a single SOP/manual document by slug.
//
// @route GET /api/v1/wiki/:slug
func (h *Handler) GetWikiPage(c *gin.Context) {
	var page models.WikiPage
	if err := h.db.First(&page, "slug = ?", c.Param("slug")).Error; err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "wiki page not found"})
		return
	}
	c.JSON(http.StatusOK, page)
}

// ListVehicles returns the structured Vehicle Database / IFF table.
//
// @route GET /api/v1/vehicle-database
func (h *Handler) ListVehicles(c *gin.Context) {
	var vehicles []models.VehicleDatabase
	if err := h.db.Order("name ASC").Find(&vehicles).Error; err != nil {
		logHandlerErr(c, "ListVehicles", http.StatusInternalServerError, "could not list vehicles")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not list vehicles"})
		return
	}
	c.JSON(http.StatusOK, gin.H{"data": vehicles})
}

// wikiInput is the body for authoring a wiki page (admin).
type wikiInput struct {
	Category string `json:"category" binding:"required"`
	Title    string `json:"title" binding:"required"`
	Icon     string `json:"icon"`
	BodyMD   string `json:"body_md" binding:"required"`
	NavOrder int    `json:"nav_order"`
}

// UpsertWikiPage creates or replaces a wiki page at the given slug (admin only).
//
// @route PUT /api/v1/wiki/:slug
func (h *Handler) UpsertWikiPage(c *gin.Context) {
	var in wikiInput
	if err := c.ShouldBindJSON(&in); err != nil {
		logHandlerErr(c, "UpsertWikiPage", http.StatusBadRequest, "category, title and body_md are required")
		c.JSON(http.StatusBadRequest, gin.H{"error": "category, title and body_md are required"})
		return
	}
	editor := middleware.DiscordID(c)
	page := models.WikiPage{
		Slug:      c.Param("slug"),
		Category:  in.Category,
		Title:     in.Title,
		Icon:      in.Icon,
		BodyMD:    in.BodyMD,
		NavOrder:  in.NavOrder,
		UpdatedBy: &editor,
		UpdatedAt: time.Now(),
	}
	if err := h.db.Clauses(clause.OnConflict{
		Columns: []clause.Column{{Name: "slug"}},
		DoUpdates: clause.AssignmentColumns([]string{
			"category", "title", "icon", "body_md", "nav_order", "updated_by", "updated_at",
		}),
	}).Create(&page).Error; err != nil {
		logHandlerErr(c, "UpsertWikiPage", http.StatusInternalServerError, "could not save wiki page")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not save wiki page"})
		return
	}
	// Reload to reflect generated/normalized fields.
	if err := h.db.First(&page, "slug = ?", page.Slug).Error; err != nil {
		logHandlerErr(c, "UpsertWikiPage", http.StatusInternalServerError, "reload after upsert failed")
		c.JSON(http.StatusInternalServerError, gin.H{"error": "could not load wiki page"})
		return
	}
	c.JSON(http.StatusOK, page)
}

// loadCurrentModpack is shared by the dashboard and modpack endpoints.
func (h *Handler) loadCurrentModpack() (*modpackDTO, error) {
	var mp models.Modpack
	err := h.db.First(&mp, "is_current = ?", true).Error
	if err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, nil
		}
		return nil, err
	}
	return h.withMods(mp), nil
}

File: apps/website/src/handlers/wiki.rs
//! Wiki + vehicle read/author handlers — Rust port of `handlers/wiki.go`.

use axum::extract::rejection::JsonRejection;
use axum::extract::{Path, State};
use axum::response::Json;
use serde::Deserialize;
use serde_json::{Value, json};

use crate::error::ApiError;
use crate::middleware::{AdminUser, AuthUser};
use crate::models::{VehicleDatabase, WikiPage};
use crate::state::AppState;

/// `GET /api/v1/wiki` — SOP nav list.
///
/// @route GET /api/v1/wiki
pub async fn list_wiki(
    State(state): State<AppState>,
    _u: AuthUser,
) -> Result<Json<Value>, ApiError> {
    let pages: Vec<WikiPage> =
        sqlx::query_as("SELECT * FROM wiki_pages ORDER BY nav_order ASC, title ASC")
            .fetch_all(&state.pool)
            .await?;
    Ok(Json(json!({ "data": pages })))
}

/// `GET /api/v1/wiki/:slug` — one SOP document.
///
/// @route GET /api/v1/wiki/:slug
pub async fn get_wiki_page(
    State(state): State<AppState>,
    _u: AuthUser,
    Path(slug): Path<String>,
) -> Result<Json<WikiPage>, ApiError> {
    let page: Option<WikiPage> = sqlx::query_as("SELECT * FROM wiki_pages WHERE slug = $1")
        .bind(&slug)
        .fetch_optional(&state.pool)
        .await?;
    page.map(Json)
        .ok_or_else(|| ApiError::not_found("wiki page not found"))
}

/// `GET /api/v1/vehicle-database` — the Vehicle Database / IFF table.
///
/// @route GET /api/v1/vehicle-database
pub async fn list_vehicles(
    State(state): State<AppState>,
    _u: AuthUser,
) -> Result<Json<Value>, ApiError> {
    let vehicles: Vec<VehicleDatabase> =
        sqlx::query_as("SELECT * FROM vehicle_databases ORDER BY name ASC")
            .fetch_all(&state.pool)
            .await?;
    Ok(Json(json!({ "data": vehicles })))
}

/// Body for authoring a wiki page (admin).
#[derive(Debug, Deserialize)]
pub struct WikiInput {
    #[serde(default)]
    category: String,
    #[serde(default)]
    title: String,
    #[serde(default)]
    icon: String,
    #[serde(default)]
    body_md: String,
    #[serde(default)]
    nav_order: i64,
}

/// `PUT /api/v1/wiki/:slug` — create or replace a wiki page (admin).
///
/// @route PUT /api/v1/wiki/:slug
pub async fn upsert_wiki_page(
    State(state): State<AppState>,
    admin: AdminUser,
    Path(slug): Path<String>,
    body: Result<Json<WikiInput>, JsonRejection>,
) -> Result<Json<WikiPage>, ApiError> {
    let Json(input) =
        body.map_err(|_| ApiError::bad_request("category, title and body_md are required"))?;
    if input.category.is_empty() || input.title.is_empty() || input.body_md.is_empty() {
        return Err(ApiError::bad_request(
            "category, title and body_md are required",
        ));
    }
    sqlx::query(
        "INSERT INTO wiki_pages (slug, category, title, icon, body_md, nav_order, updated_by, updated_at) \
         VALUES ($1, $2, $3, $4, $5, $6, $7, now()) \
         ON CONFLICT (slug) DO UPDATE SET category = EXCLUDED.category, title = EXCLUDED.title, \
          icon = EXCLUDED.icon, body_md = EXCLUDED.body_md, nav_order = EXCLUDED.nav_order, \
          updated_by = EXCLUDED.updated_by, updated_at = now()",
    )
    .bind(&slug)
    .bind(&input.category)
    .bind(&input.title)
    .bind(&input.icon)
    .bind(&input.body_md)
    .bind(input.nav_order)
    .bind(&admin.0.discord_id)
    .execute(&state.pool)
    .await?;

    let page: WikiPage = sqlx::query_as("SELECT * FROM wiki_pages WHERE slug = $1")
        .bind(&slug)
        .fetch_one(&state.pool)
        .await?;
    Ok(Json(page))
}
